lib/testutils/mini_runner.go

```diff
@@ -0,0 +1,137 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package testutils
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/stats"
+)
+
+// Ensure mock implementations conform to the interfaces.
+var _ lib.Runner = &MiniRunner{}
+var _ lib.VU = &MiniRunnerVU{}
+
+// MiniRunner partially implements the lib.Runner interface, but instead of
+// using a real JS runtime, it allows us to directly specify the options and
+// functions with Go code.
+type MiniRunner struct {
+	Fn         func(ctx context.Context, out chan<- stats.SampleContainer) error
+	SetupFn    func(ctx context.Context, out chan<- stats.SampleContainer) ([]byte, error)
+	TeardownFn func(ctx context.Context, out chan<- stats.SampleContainer) error
+
+	setupData []byte
+
+	CurrentVU int64
```

Comment:
- Does this have to be public?
- Why is it called "current"? It's used as a counter for the VU ID, but I'm confused by "current", as it implies this is the currently active VU or something to that effect. Wouldn't "last" be more accurate, or even "count" since it's probably best left private, unless it has an external use case I'm missing?

Reply:
Maybe calling it `NextVUID` will be better ... also maybe `nextVUID`

Reply:
Regarding public - probably yes. In general, for testing harnesses, helpers and mock objects, I err on the side of making things public, since the actual tests often need to know about the internals of the mock object.

Regarding the name, `NextVUID` seems better.

Reply:
Thinking about it, I even want to make `setupData` public...

Reply:
OK, I see how it can be useful, but maybe YAGNI? :D I'm fine with both being exported if you expect to use them in the immediate future, though.

And :+1: for `NextVUID`.

Reply:
I'm not adding new functionality, I'm just not hiding already existing functionality, so YAGNI doesn't really apply...

One of the biggest reasons we have have public/private (or, in Go terms, unexported/exported) properties and methods is that users of a certain abstraction shouldn't need to know the implementation details of the thing they are using. But in the case of test mock objects, the implementation details are usually half the point :smile: . A good example is the httpmultibin - all of its properties are public, which has often been handy...

Reply:
Well, visibility is one thing, but mutability is another. What happens if the test modifies `MiniRunner.NextVUID`? Is it allowed to? In this case it's pretty clear, `MiniRunner` is simple enough, and as you say, the test and `MiniRunner` should be intimate ;) since it's part of `testutils`. Otherwise making properties public usually brings up other worries and complexity that can be avoided by defaulting to private.

So I'm fine with this approach here, but we should probably avoid it for `js.Runner` tests or higher-level tests in general, which should be more black-boxy.

Reply:
But that's the thing about mock objects - tests are expected to have full knowledge of their internals, so yes, pretty intimate :sweat_smile: 

And yeah, it very much depends on the test - for some tests you'd just need the public API, so you don't need to tinker with the internals. But for those you could just as easily use the actual JS `Runner`.